src/GNDStk/tree/test/write.test.cpp

```diff
@@ -0,0 +1,314 @@
+                     "000000000047compression": "diagonal"
+                  },
+                  "000000000048values": {
+                     "000000000049attributes": {
+                        "000000000050body": "0.015 0 0 0 4.5e-5 0.015 3e-2 0 0 0 1.35e-4 0.015 2e-2 0 0 0 1.5e-3 0.012 5e-2 0 0 0 1.875e-3 6e-2 5e-2 0 0 0 1.05e-4 0.015 0.1 0 0 0 6e-4 0.012 0.1 0 0 0 2.25e-4 0.012 0.2 0 0 0 5.25e-3 0.012 0.2 0 0 0 3.45e-3 0.012 0.3 0 0 0 4.5e-4 0.012 0.3 0 0 0 3e-3 0.012 0.4 0 0 0 9e-3 0.012 0.4 0 0 0 1.425e-3 0.012"
+                     }
+                  }
+               }
+            }
+         }
+      }
+   }
+})***";
+
+
+
+// -----------------------------------------------------------------------------
+// SCENARIO
+// -----------------------------------------------------------------------------
+
+// fixme As elsewhere, we don't yet have anything of substance for HDF5.
+
+SCENARIO("Testing GNDStk tree write() and operator<<") {
+   gnds::Tree<> tree;
+
+   // ------------------------
+   // Empty tree
+   // ------------------------
+
+   GIVEN("An empty gnds::tree object") {
+      // For now I'll just write to an output stream, as I'm not sure how our
+      // testing system is supposed to work when our intention is to actually
+      // write *files*.
+
+      // format::tree
+      WHEN("We write() or << the empty tree in format::tree format") {
+         {
+            // write()
+            std::ostringstream oss;
+            tree.write(oss, gnds::format::tree);
+            REQUIRE(oss.str() == string_empty_tree);
+         } {
+            // write(), using format::null (which defaults to tree)
+            std::ostringstream oss;
+            tree.write(oss, gnds::format::null);
+            REQUIRE(oss.str() == string_empty_tree);
+         } {
+            // <<
+            std::ostringstream oss;
+            oss << tree;
+            REQUIRE(oss.str() == string_empty_tree);
+         }
+      }
+
+      // format::xml
+      WHEN("We write() the empty tree in format::xml format") {
+         std::ostringstream oss;
+         tree.write(oss, gnds::format::xml);
+         REQUIRE(oss.str() == string_empty_xml);
+      }
+
+      // format::json
+      WHEN("We write() the empty tree in format::json format") {
+         std::ostringstream oss;
+         tree.write(oss, gnds::format::json);
+         REQUIRE(oss.str() == string_empty_json);
+      }
+   }
+
+   // ------------------------
+   // Real tree, from one of
+   // our sample GNDS files.
+   // ------------------------
+
+   GIVEN("A real gnds::tree, read from a file") {
+      tree.read("n-069_Tm_170-covar.xml");
+      REQUIRE(!tree.empty());
+
+      // format::tree
+      WHEN("We write() or << the tree in format::tree format") {
+         {
+            // write()
+            std::ostringstream oss;
+            tree.write(oss, gnds::format::tree);
+            REQUIRE(oss.str() == string_real_tree);
```

Comment:
The abillity to write the raw tree structure is something that I put in, because I think it's good as a debugging tool. (There's no need for the corresponding read, which is why I don't have it.) With a raw tree writer capability, I can deal with situations where, for instance, somebody presents an .xml and .json that they assert (correctly or not) are supposed to represent equivalent GNDS hierarchies. A quick check could involve reading both, dumping out the raw tree structure, and seeing if they're the same. There are certainly other ways to do such a check, e.g. by reading one of the files, converting to the other format, then writing it out in the other format and comparing with the file that was originally in the other format. Really, this provides a capability that was easy to write, and gives us a simple and convenient tool to do some additional checking above and beyond what we can already do. Note also that my tree dump capability is more predictable. With json and xml writes (via our two external libraries), I've found that whitespace treatment can be unpredictable; this can confound comparisons done in that manner. (Incidentally, doing a comparison as I've described has a bigger confounding factor, in particular that GNDS doesn't really have an ordering concept, so child nodes or metadata could be shuffled, making two equivalent hierarchies appear different when they're actually the same. One planned capability on my already long to-do list is something that would "normalize" (there's that word again) a given hierarchy so that order-agnostic comparisons can be made.)